### ISSUE TITLE
Automated IDs for requirements based on identifier of requirement

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-ogc#663]
from https://github.com/metanorma/metanorma-ogc/issues/659

There's a few new requirements now, and I am separating them out, because I'm not as enthusiastic about all of them:

* Requirements get hover anchor IDs
* Requirements get automatic ID based on the requirement ID, and not a GUID (unless the editor has supplied an explicit anchor)
* Test table gets automatic anchor—by which I think you mean the individual tests in Annex A
* Implicit from the foregoing, automatic ID based on the test ID and not a GUID (unless the editor has supplied an explicit anchor)

I can do hover anchors, and I'll attempt it now, although the tabular rendering may make the sidebar § look ugly. 

I am not as enthusiastic about anchors based on the IDs: the IDs will need to remove characters disallowed in XML anchors, so they won't be identical to supplied IDs, and they will be lengthy for anchors. Plus, intervening in that way may be more complicated than I'd like: I'd have to investigate the Modspec-specific code, and get endorsement that OGC is OK with semantically-based automatic hyperlinks for them.

**Comment by opoudjis** [comment]
From @jerstlouis:

Thanks a lot @opoudjis !

> the IDs will need to remove characters disallowed in XML anchors, so they won't be identical to supplied IDs,

Are forward slashes `/` allowed? Because requirement identifiers usually only use alphanumeric characters, underscores and slashes.

> and they will be lengthy for anchors

I really think this is the single most valuable thing to be able to directly link to a particular requirement, using an anchor corresponding to that requirement ID.

>  get endorsement that OGC is OK with semantically-based automatic hyperlinks for them

Cool, I hope they like idea, but it sounds very much aligned with what metanorma is trying to do to generate as much as possible from a single source of truth. One less place that editors would need to manually ensure consistency between requirement identifiers and the anchors.

> which I think you mean the individual tests in Annex A

That is what I meant, yes.

> I can do hover anchors, and I'll attempt it now, although the tabular rendering may make the sidebar § look ugly.

If the § hover anchors is not suitable somewhere (e.g., in the TOC), simply the Section / Requirement X / Test X header text being right click / _Copy link_ -able is just as good. I personally prefer the thing just being right click - _Copy link_ -able without the hover thing but that's just me, and I got used to the § hover.

> So I'd like to separate that aspect to a separate ticket, and defer it.

That makes sense, we've gone quite far from my original simple question :)

**Comment by opoudjis** [comment]
This is in progress. There are hover links but no § yet, and I need to disable the visited link colouring if the anchors are clicked.

**Comment by opoudjis** [comment]
Abstracting method to detect GUID anchors in Metanorma.

**Comment by opoudjis** [comment]
> Are forward slashes `/` allowed? Because requirement identifiers usually only use alphanumeric characters, underscores and slashes.

No. That's the source of my hesitation.

This has ended up being easy to implement, but I'm going to wait on approval from OGC before publishing this as an update. The change is: if no explicit identifier has been provided for a Modspec requirement (so that Metanorma has supplied a GUID), this requirement will get an ID in postprocessing that uses the first identifier it contains—but sanitised so that it can be used as an anchor.

This is what the XML looks like:

````
      <sections>
        <requirement id="http___www.opengis.net_spec_waterml_2.0_req_xsd-xml-rules" model="ogc" type="class">
          <identifier>http://www.opengis.net/spec/waterml/2.0/req/xsd-xml-rules</identifier>
        </requirement>
      </sections>
````

So given an identifier `http://www.opengis.net/spec/waterml/2.0/req/xsd-xml-rules`, you get the anchor `http___www.opengis.net_spec_waterml_2.0_req_xsd-xml-rules`: Metanorma replaces all characters illegal in an anchor with underscore.

**Comment by opoudjis** [comment]
I am attaching what the sample document looks like now with these changes. As said, not incorporating these updates into release without approval, but these changes are more benign than I had feared.

A few requirements still have GUIDs, because they have not been assigned identifiers; e.g. _49f0afd3-40f3-2900-00e4-7ad28a59a621 in the following (showing Semantic Metanorma XML):

````xml
<clause id="_conformance_class_basic_vector_features_styling" obligation="normative">
<title>Conformance Class “Basic Vector Features Styling”</title>
<requirement id="_49f0afd3-40f3-2900-00e4-7ad28a59a621" model="ogc" type="conformanceclass"> <identifier>http://www.opengis.net/spec/symbology-1/2.0/conf/vector</identifier><subject><xref target="rc-core">Requirements Class “Basic Vector Features Styling”</xref></subject><classification><tag>Target Type</tag><value>Encoding</value></classification><classification><tag>Target Type</tag><value>Rendering engine</value></classification>
</requirement>

<requirement id="_conf_vector_stroke" model="ogc" type="abstracttest"> <identifier>/conf/vector/stroke</identifier><classification><tag>target</tag><value>/req/vector/stroke</value></classification><component class="test-purpose"><p id="_1687509c-da81-c534-658c-869b46731075">Validate the encoding of a symbolizer stroke</p>
</component><component class="test-method"><p id="_a2a46cf2-bd98-1772-e87c-2799115c85a0">Given:</p>

<p id="_ce1fae14-f683-ae1d-0768-7b6c0644aa61">When:</p>

<p id="_1bc91431-e742-6ba5-5839-d05c197e934d">Then:</p>
</component>
</requirement>
</clause>
````

[18-067r4 3.html.zip](https://github.com/metanorma/metanorma-ogc/files/15188936/18-067r4.3.html.zip)

**Comment by jerstlouis** [comment]
@opoudjis Thank you, this is really great stuff.

One comment about those §: in this last version, the sections are right-click "Copy link"-able and get highlighted themselves, in addition to the § on the left side. That seems a bit redundant? With plain ASCIIDoctor before you would have to move the mouse over to the § to copy the link. If the header itself highlights and allows to copy the link, then we don't really need the § ?

Myself I prefer the section itself and no § (faster -- no need to move the mouse over to the §), but @ghobona and @gbuehler can chime in on their preference.

**Comment by opoudjis** [comment]
@ghobona @gbuehler Reminder that I need feedback from you on this in order to incorporate into release. We are meeting in a week for our regular catchup, right?

**Comment by opoudjis** [comment]
While we don't *need* the §, I think it's harmless to include it as a indicator that there is a link there. The hover anchor is a widespread convention.

> So maybe just showing the § as the "hovering" indication (no background highlight in this case), but not having to move the mouse over to the § to copy the link, would be ideal.

That makes sense to me, I'll ensure it.

**Comment by opoudjis** [comment]
Both the § and the title are hyperlinked, but you do not need to hover to the § to hyperlink, clicking on the title is enough.

**Comment by jerstlouis** [comment]
@opoudjis 

> but you do not need to hover to the § to hyperlink, clicking on the title is enough.

That's a good thing! :)

What was annoying me specifically was:
- They were two separate anchors, where one or the other highlights rather than both highlighting together, even though the actual link was the same
- The big sections getting a full background highlight, which does not feel as clean as the old ASCII docs where only the § appears

**Comment by opoudjis** [comment]
For final approval...

[18-067r4 4.html.zip](https://github.com/metanorma/metanorma-ogc/files/15468638/18-067r4.4.html.zip)

**Comment by jerstlouis** [comment]
@opoudjis 

The section link works great in this build, but somehow the requirements are not there, and the conformance class / abstract tests do not have links like they did in the earlier version?

I don't know if it's because you're using a smaller extract of the source spec, or some regression was introduced...

**Comment by opoudjis** [comment]
It was a smaller extract: for some reason, it was taking forever to compile.

OK, closing this ticket.

**Comment by opoudjis** [comment]
Document in modspec
